src/libraries/Microsoft.PowerFx.Core/Types/LazyTypeProvider.cs

```diff
@@ -0,0 +1,84 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Text;
+using Microsoft.PowerFx.Core.Types;
+using Microsoft.PowerFx.Core.Utils;
+using Microsoft.PowerFx.Types;
+
+namespace Microsoft.PowerFx.Types
+{
+    public sealed class LazyTypeProvider
+    {
+        public delegate FormulaType FieldTypeGetter();
+
+        private readonly ImmutableDictionary<DName, FieldTypeGetter> _fieldTypeGetters;
+        private readonly Dictionary<DName, DType> _expandedFields = new Dictionary<DName, DType>();
+
+        public readonly ILazyTypeMetadata LazyTypeMetadata;
+
+        public LazyTypeProvider(ILazyTypeMetadata metadata, Dictionary<DName, FieldTypeGetter> fieldTypeGetters)
+        {            
+            LazyTypeMetadata = metadata;
+            _fieldTypeGetters = fieldTypeGetters.ToImmutableDictionary();
+        }
+
+        private LazyTypeProvider(ILazyTypeMetadata metadata, ImmutableDictionary<DName, FieldTypeGetter> fieldTypeGetters)
+        {            
+            LazyTypeMetadata = metadata;
+            _fieldTypeGetters = fieldTypeGetters;
+        }
+
+        internal bool TryGetFieldType(DName name, out DType type)
+        {
+            if (_expandedFields.TryGetValue(name, out type))
```

Comment:
We cache field access that has already been expanded, otherwise, we have a callback to get the type of each field in the type